### ISSUE TITLE
codenvy/cli connect to docker on TLS

### ISSUE THREAD

**Issue by wernight** [codenvy/codenvy#1954]
**Codenvy version:** 5.5.0

## Issue

By default `codenvy/cli` will open Docker port 23750 and **any user having access to it will have `root`-like** access to this machine. This means anyone on the internet if the port is exposed or anyone on the network if not. [Docker has a mechanism to secure socket connections](https://docs.docker.com/engine/security/https/) and `codenvy/cli` may **use it**, or offer to use it during install, and **warn** other wise.

## Background:

For Codenvy on-premise users, the simplest (and likely most common) path is using `docker ... codenvy/cli start` on a single machine (especially it's limited to 3 users most people won't need more than one and don't have more than one). It also means that this machine is likely to have rights to deploy websites to production.

It's say that the first and simplest step is to warn, as blocking that port may be seen as secure enough for most users. The second may be to encrypt by default.

See also https://github.com/codenvy/codenvy/issues/1944#issuecomment-287602237

**Comment by eivantsov** [comment]
Just a side note, codenvy/cli does not open any ports. Port 23750 should be opened just for local connections, not for inbound ones from outside the node network.

**Comment by TylerJewell** [comment]
@wernight - I completely agree that connections to Docker should be managed in a secure way. But let's make sure that the scope of the issue is what we are all expecting it to be.  I am getting tripped up on some of your definitions, which may be different in your head than they are in Eugene or mine.

First, the `codenvy/cli` image is just a launcher image. It's stateless and only runs for a limited period of time. It does connect to the Docker daemon in a couple situations:
1. To verify that access to the daemon is available either over unix socket or `DOCKER_HOST`. 
2. To run various utility impages like `alpine` or `eclipse/che-ip` to generate values.
3. To run our puppet image, which is used to convert a `codenvy.env` to `docker-compose-container.yml`.
4. And then to run `docker/compose` image that takes the generate compose file as input.

Second, the `docker/compose` image runs a Docker Compose service, which is the runtime capabilities of Codenvy. This uses containers for Swarm, ZooKeeper, Socat, Postgresql, Codenvy, HAProxy, and Nginx to create a runtime infrastructure.

Third, the runtime infrastructure also needs to have access to the Docker daemon, but through Swarm.  In Eclipse Che, it is directly to the daemon itself.  However, the runtime infrastructure can optionally be configured with SSL.  And I believe (Eugene will keep me honest here) that the SSL configuration is for all communications in between Codenvy nodes AND also for Codenvy nodes to the Docker Daemon through swarm over TLS.  

So, this comes back to the question of whether it really matters for the CLI to have TLS connectivity built into it, or if the runtime SSL connectivity for the runtime infrastructure is suitable enough?

**Comment by wernight** [comment]
I imagined CLI as doing that setup that's why I mentioned it.

Doesn't really matter to me which part is opening that port; what it should do is only allow Che system to connect to it, and even from within a workspace it shouldn't be allowed by default (would be good to have the option though).

**Comment by eivantsov** [comment]
It is you and only you who opens that port. Make it accessible only for swarm container.

**Comment by wernight** [comment]
I agree it's possible to create a subnet or set up iptable rules.

The issue here is that without warning, running `codenvy/cli` will open a port and if the server doesn't have a firewall (which is actually pretty common), then you've just opened a huge backdoor without realizing it. Another issue you've opened up is that any user on any workspace has `root` on your system by default, and any user on your machine, for example any PHP or Python script, has also `root` on your system.

**Comment by TylerJewell** [comment]
Werner I am struggling to understand your responses.  I feel like a dog chasing the tail a little.  

The original issue is whether tls is needed from the cli.  I do not believe it is needed as the production services all use tls and the cli is a temporal object.

I don't understand how the conversation shift to ports being opened or workspace permissions.

**Comment by wernight** [comment]
Steps, supposing `10.10.10.10` is your machine's public IP:

  1. On host: `DOCKER_HOST=tcp://10.10.10.10:23750 docker ps` --> `Cannot connect to the Docker daemon at tcp://localhost:2375. Is the docker daemon running?` (that's good)
  2. On host: `docker run --rm ... codenvy/cli start`
  3. On any machine or from within a workspace*: `DOCKER_HOST=tcp://10.10.10.10:23750 docker ps` --> Connects and you could actually do harm (I hope I don't need to give any example here).

*: Supposing you have a standard default distro, meaning you don't have an explicit iptable rule blocking this.

You say that "production services all use tls" (I suppose you refer here that `codenvy/cli` spawn these "production services"). However it seems that I can connect without having that specific certificate.

**Comment by TylerJewell** [comment]
Did you set up the SSL certificate in your `codenvy.env` file before starting Codenvy?

I would - yes - absolutely expect that the default configuration without setting up SSL would expose that risk that you outlined.  Once nice thing is that we are exploring some migrations to replace nginx with traefik and there are some techniques where we can auto generate certificates to setup SSL for the end user without them having to take actions.

**Comment by wernight** [comment]
I did not, but I do expect one of (in order of preference):

 1. Codenvy to generate on for me (may be using Let's Encrypt if necessary but I think a private one is even better)
 2. Codenvy to give me a big fat warning.
 3. Codenvy documentation to give a big warning that people should set that field in `codenvy.env` as part of the set up steps (i.e. not hidden below in the read-me but prominently as part of required steps).

**Comment by TylerJewell** [comment]
@benoitf was just looking at what was needed for us to generate a TLS cert with Let's Encrypt recently.  I am not sure if he implemented it yet.  Could you @benoitf comment briefly on pre-reqs that need to be upheld before we could start to do this?

I agree with your other points.  We are not emphasizing this enough in the startup sequence or in the docs on the essentiality of this for production purposes.

**Comment by wernight** [comment]
I believe Let's Encrypt isn't needed here because basically you'd want to allow only Codenvy to access it and you may give codenvy the private certificate that allows access along with the trusted public certificate of the private TLS certificate. Basically *Let's Encrypt* seems required only to have *public* trust, but here we need to trust only a specific client by default: Codenvy itself.

**Comment by garagatyi** [comment]
To secure docker there is no need to use public trust. Auto generated keys would be fine.
But in this case I think we just should not publish ports by default. Since all codenvy containers in one network it is not needed.

**Comment by TylerJewell** [comment]
Turns out that our engineers have been exploring a number of scenarios. There are multiple avenues that they are trying for.  It's going to require some additional research.  Nothing is simple or obvious, but they are confident that something can be done.

**Comment by riuvshin** [comment]
@TylerJewell it seems you are talking about https support for codenvy, but actual topic is about docker tls security which is different.

**Comment by riuvshin** [comment]
@skryzhny @garagatyi as i understand we just need to make exposing of socat port 23750 optional and disabled by default, right?

**Comment by garagatyi** [comment]
Yes you are right. We should not expose 23750, 23751 ports by default.

**Comment by riuvshin** [comment]
so i did some investigation on this and figured that we can't get rid of exposing 23750 port in socat container because it will break all in one (default) functionality where we using same instance for running workspaces. However we don't need it when we using external nodes for running workspaces (production mode) in this case we don't run workspaces on master node so I can make this expose configurable but enabled by default, which will allow to disable expose in production mode. WDYT guys?

**Comment by skryzhny** [comment]
I'm not really happy with solution you proposes.
We need to get rid of this exposure, as it really bad for users who's trying codenvy on non-VM.
Maybe we need add logic that translates internal 'socat' address to external one when presenting to user's browser. So we can get rid of exposure at all.
Or somehow, probably with `extra-hosts` allow access to socat port published to 127.0.0.1 so it isn't accessible from outside of a host. However second option requires use of DNS name, not IP.
We discussed such approach even for external nodes, as sometime WorkspaceNodes may have different DNS/IP when accessed internally (codenvy software) and externally (user's browser). It would allow internally communicate with secured internal net while presenting to user externally accessible part of infrastructure.

**Comment by riuvshin** [comment]
@skryzhny I have no idea how to add that kind of "logic" you want, so here is my pr which just improves what we have now: https://github.com/codenvy/codenvy/pull/1990
But if you believe that this still not enough you can propose your changes on this as you have vision for that.

**Comment by wernight** [comment]
@riuvshin I'm not sure what that "production mode" means. A user just running `docker run codenvy/cli start` should have "production mode" but there I think it falls into the case where this port *needs* to remain open (per [your comment](https://github.com/codenvy/codenvy/issues/1954#issuecomment-289442864)).

IMO network restriction is good not a real fix for the security risk. For example by default only `root` and run docker containers because Docker gives almost root power. Having it listen only to localhost would still allow any local process, running as any user, capable of sending data to 127.0.0.1:23750 and get root-like power.

Requiring a specific client key (which I think docker supports) sounds like a better initial security. Limiting access to that port sounds like a secondary important feature and just documenting it could allow most people to set up firewall rules.

**Comment by riuvshin** [comment]
@wernight by default when user run codenvy in starts  in single node configuration which means that workspaces will be scheduled to start on same host. This is not recommended for production because many workspaces with their activities may cause perf issues for codenvy master. But this kind of configuration is much easier for initial set up and evaluation.
For production mode remote nodes should be added to codenvy system and codenvy master will schedule workspace creation on them via swarm.
So in case where you have remote nodes you can just disable expose of socat port which is needed only for all-in-one set up.

**Comment by garagatyi** [comment]
@riuvshin we can set property that overrides nodes hostname. It can be used in AIO usecase only, but it will help in this situation. If a user change default nodes list of Swarm he would have to comment that property, change list of nodes.

**Comment by eivantsov** [comment]
@garagatyi yes, we discussed that - a property that provides a custom hostname that is used to display preview URLs, while swarm will connect to its node using scat alias name.

**Comment by skryzhny** [comment]
We were tried and found that swarm can't be configured to listen on http while connecting to docker/socat on https.
So we have two options:
1. Use pairs of internal/external names for nodes. Internal names for configuring swarm and external names for presenting links to user and starting agents in workspaces, etc.
2. Make codenvy able to talk to swarm via https.

**Comment by TylerJewell** [comment]
@skryzhny @benoitf - the safest long term approach would be your #1 - which is that we create a formal separation between external names for nodes / servers / ports / swarm and their internal representations.  By doing this mapping, we can then apply appropriate security layers.  This is somewhat similar to the work done by Red Hat in consolidating traffic through a single port.

**Comment by TylerJewell** [comment]
Will not support this feature improvement. Instead we will port to run on top of k8s, which will manage this for us.
